practice/tic-tac-toe/coords_board.py

```diff
@@ -0,0 +1,166 @@
+        'O'
+
+        >>> a = CoordsTTTBoard()
+        >>> a._token_coords = [
+        ... (0, 0, 'X'), (2, 0, 'O'),
+        ... (1, 1, 'X'),
+        ... (2, 2, 'X')
+        ... ]
+        >>> a.calc_winner()
+        'X'
+        """
+        win_coords = [
+            [item[2] for item in self._token_coords if item[0] == 0],
+            [item[2] for item in self._token_coords if item[0] == 1],
+            [item[2] for item in self._token_coords if item[0] == 2],
+            [item[2] for item in self._token_coords if item[1] == 0],
+            [item[2] for item in self._token_coords if item[1] == 1],
+            [item[2] for item in self._token_coords if item[1] == 0],
+            [item[2] for item in self._token_coords if item[1] == 1],
+            [item[2] for item in self._token_coords if item[1] == 2],
+            [item[2] for item in self._token_coords if
+                (item[0] == 0 and item[1] == 2) or
+                (item[0] == 1 and item[1] == 1) or
+                (item[0] == 2 and item[1] == 0)],
+            [item[2] for item in self._token_coords if
+                (item[0] == 0 and item[1] == 0) or
+                (item[0] == 1 and item[1] == 1) or
+                (item[0] == 2 and item[1] == 2)]
+        ]
+        if ['X', 'X', 'X'] in win_coords:
+            return 'X'
+        elif ['O', 'O', 'O'] in win_coords:
+            return 'O'
+        else:
+            return None
+
+    def __str__(self):
+        r"""Returns a pretty-printed picture of the Board .
+
+        >>> a = CoordsTTTBoard()
+        >>> a._token_coords= [
+        ... (0, 0, 'X'), (1, 0, 'X'), (2, 0, 'O'),
+        ... (0, 1, 'O'), (1, 1, 'O'), (2, 1, 'X'),
+        ... (0, 2, 'X'), (2, 2, 'X')
+        ... ]
+        >>> a.__str__()
+        'X|X|O\nO|O|X\nX| |X\n'
+        """
+        cell00 = ' '
+        cell01 = ' '
+        cell02 = ' '
+        cell10 = ' '
+        cell11 = ' '
+        cell12 = ' '
+        cell20 = ' '
+        cell21 = ' '
+        cell22 = ' '
+        for coord in self._token_coords:
+            if coord[0] == 0 and coord[1] == 0:
+                cell00 = coord[2]
+            if coord[0] == 0 and coord[1] == 1:
+                cell01 = coord[2]
+            if coord[0] == 0 and coord[1] == 2:
+                cell02 = coord[2]
+            if coord[0] == 1 and coord[1] == 0:
+                cell10 = coord[2]
+            if coord[0] == 1 and coord[1] == 1:
+                cell11 = coord[2]
+            if coord[0] == 1 and coord[1] == 2:
+                cell12 = coord[2]
+            if coord[0] == 2 and coord[1] == 0:
+                cell20 = coord[2]
+            if coord[0] == 2 and coord[1] == 1:
+                cell21 = coord[2]
+            if coord[0] == 2 and coord[1] == 2:
+                cell22 = coord[2]
+        return cell00 + '|' + cell10 + '|' + cell20 + '\n' +\
+            cell01 + '|' + cell11 + '|' + cell21 + '\n' +\
+            cell02 + '|' + cell12 + '|' + cell22 + '\n'
```

Comment:
There is a really nice parallel between `str.join()` and what the goal of this function is. Try using it instead of manually interpolating.